java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/SearchDictionaryEncoder.java

```diff
@@ -0,0 +1,111 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.dictionary;
+
+import org.apache.arrow.algorithm.search.VectorSearcher;
+import org.apache.arrow.algorithm.sort.VectorValueComparator;
+import org.apache.arrow.vector.BaseIntVector;
+import org.apache.arrow.vector.BitVectorHelper;
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * Dictionary encoder based on searching.
+ * @param <E> encoded vector type.
+ * @param <D> decoded vector type, which is also the dictionary type.
+ */
+public class SearchDictionaryEncoder<E extends BaseIntVector, D extends ValueVector> {
+
+  /**
+   * The dictionary for encoding/decoding.
+   * It must be sorted.
+   */
+  protected final D dictionary;
+
+  /**
+   * The criteria by which the dictionary is sorted.
+   */
+  protected final VectorValueComparator<D> comparator;
+
+  /**
+   * A flag indicating if null should be encoded.
+   */
+  protected final boolean encodeNull;
+
+  /**
+   * Constructs a dictionary encoder.
+   * @param dictionary the dictionary. It must be in sorted order.
+   * @param comparator the criteria for sorting.
+   */
+  public SearchDictionaryEncoder(D dictionary, VectorValueComparator<D> comparator) {
+    this(dictionary, comparator, false);
+  }
+
+  /**
+   * Constructs a dictionary encoder.
+   * @param dictionary the dictionary. It must be in sorted order.
+   * @param comparator the criteria for sorting.
+   * @param encodeNull if null should be encoded.
+   */
+  public SearchDictionaryEncoder(D dictionary, VectorValueComparator<D> comparator, boolean encodeNull) {
+    this.dictionary = dictionary;
+    this.comparator = comparator;
+    this.encodeNull = encodeNull;
+  }
+
+  /**
+   * Encodes an input vector by binary search.
+   * So the algorithm takes O(nlogm) time, where n is the length of the input vector,
+   * and m is the length of the dictionary.
+   * @param input the input vector.
+   * @param output the output vector.
+   */
+  public void encode(D input, E output) {
+    for (int i = 0; i < input.getValueCount(); i++) {
+      if (!encodeNull && input.isNull(i)) {
+        BitVectorHelper.setValidityBit(output.getValidityBuffer(), i, 0);
+        continue;
+      }
+
+      int index = VectorSearcher.binarySearch(dictionary, comparator, input, i);
+      if (index == -1) {
+        throw new IllegalStateException("The data element is not found in the dictionary");
+      }
+      output.setWithPossibleTruncate(i, index);
+    }
+    output.setValueCount(input.getValueCount());
+  }
+
+  /**
+   * Decodes an input vector. The algorithm takes O(n) time,
+   * where n is the length of the input vector.
+   * @param input the input vector.
+   * @param output the output vector.
+   */
+  public void decode(E input, D output) {
```

Comment:
is there a reason to not keep this as a static method on the existing VectorEncoderClass, if we want to have a separately allocated vector that seems like an easy change?

Reply:
It seems the existing DictionaryEncoder class does not support the encodeNull flag.

Reply:
In response to this comment, I have extracted a static utility method for the decode functionality.

Reply:
I don't think we need the decode method at all here, it seems like we can rely on the one in vector.  Per above comment, I don't understand the purpose of the encodeNull flag on decode is.

Reply:
The purpose of the encodeNull is discussed above. 
You mean we can rely on the one in the static method?